<a href="https://colab.research.google.com/github/baharahmadpoor/PINN/blob/main/Brestcancer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

to in ravesh image nist

In [82]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from sklearn.datasets import load_breast_cancer

In [52]:
bc = load_breast_cancer()
print(bc.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

:Number of Instances: 569

:Number of Attributes: 30 numeric, predictive attributes and the class

:Attribute Information:
    - radius (mean of distances from center to points on the perimeter)
    - texture (standard deviation of gray-scale values)
    - perimeter
    - area
    - smoothness (local variation in radius lengths)
    - compactness (perimeter^2 / area - 1.0)
    - concavity (severity of concave portions of the contour)
    - concave points (number of concave portions of the contour)
    - symmetry
    - fractal dimension ("coastline approximation" - 1)

    The mean, standard error, and "worst" or largest (mean of the three
    worst/largest values) of these features were computed for each image,
    resulting in 30 features.  For instance, field 0 is Mean Radius, field
    10 is Radius SE, field 20 is Worst Radius.

    - 

benign -> badkhim

malingnant -> khoshkhim

In [53]:
bc.target[500]

np.int64(1)

In [54]:
bc.target.shape

(569,)

In [55]:
bc.data[500]

array([1.504e+01, 1.674e+01, 9.873e+01, 6.894e+02, 9.883e-02, 1.364e-01,
       7.721e-02, 6.142e-02, 1.668e-01, 6.869e-02, 3.720e-01, 8.423e-01,
       2.304e+00, 3.484e+01, 4.123e-03, 1.819e-02, 1.996e-02, 1.004e-02,
       1.055e-02, 3.237e-03, 1.676e+01, 2.043e+01, 1.097e+02, 8.569e+02,
       1.135e-01, 2.176e-01, 1.856e-01, 1.018e-01, 2.177e-01, 8.549e-02])

In [56]:
bc.data.shape

(569, 30)

In [57]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(bc.data, bc.target, test_size=0.2, random_state=42)
x_train.shape, x_test.shape, y_train.shape, y_test.shape

((455, 30), (114, 30), (455,), (114,))

In [58]:
from sklearn.preprocessing import MinMaxScaler
scalar=MinMaxScaler(feature_range=(0,1))
X_train=scalar.fit_transform(x_train)
X_test=scalar.transform(x_test)

In [59]:
X_train[200]

array([0.56266223, 0.43219479, 0.56608337, 0.4194628 , 0.29430351,
       0.42264228, 0.26101218, 0.33658052, 0.3203417 , 0.11225779,
       0.11739996, 0.15753182, 0.1449371 , 0.08856589, 0.12955094,
       0.26082254, 0.06833333, 0.23129381, 0.11698832, 0.08644611,
       0.53658358, 0.53411514, 0.56890855, 0.36333863, 0.4145215 ,
       0.58511328, 0.30870607, 0.58694158, 0.32091465, 0.31510682])

In [60]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
model.fit(X_train, y_train)

GaussianNB()

In [61]:
y_pred_train = model.predict(X_train)
y_pred_test = model.predict(X_test)

In [80]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
def calculate_metrics(y_train, y_test,y_pred_train, y_pred_test):

    acc_train = accuracy_score(y_true=y_train, y_pred=y_pred_train)
    acc_test = accuracy_score(y_true=y_test, y_pred=y_pred_test)

    p_train = precision_score(y_true=y_train, y_pred=y_pred_train)
    p_test = precision_score(y_true=y_test, y_pred=y_pred_test)

    r_train = recall_score(y_true=y_train, y_pred=y_pred_train)
    r_test = recall_score(y_true=y_test, y_pred=y_pred_test)

    f1_train = f1_score(y_true=y_train, y_pred=y_pred_train)
    f1_test = f1_score(y_true=y_test, y_pred=y_pred_test)

    conf_train = confusion_matrix(y_true=y_train, y_pred=y_pred_train)
    conf_test = confusion_matrix(y_true=y_test, y_pred=y_pred_test)

    print(f"accuracy train: {acc_train}, accuracy test: {acc_test}")
    print()
    print(f"precition train: {p_train}, precition test: {p_test}")
    print()
    print(f"recall train: {r_train}, recall test: {r_test}")
    print()
    print(f"f1 train: {f1_train}, f1 test: {f1_test}")
    print()
    print(f"confusion matrix train:\n{conf_train}")
    print()
    print(f"cofusion matrix test:\n{conf_test}")
    print()

    # return acc_train, acc_test, p_train, p_test, r_train, r_test, f1_train, f1_test

In [81]:
calculate_metrics(y_train, y_test,y_pred_train, y_pred_test)

accuracy train: 0.9362637362637363, accuracy test: 0.9649122807017544

precition train: 0.9355932203389831, precition test: 0.958904109589041

recall train: 0.965034965034965, recall test: 0.9859154929577465

f1 train: 0.9500860585197934, f1 test: 0.9722222222222222

confusion matrix train:
[[150  19]
 [ 10 276]]

cofusion matrix test:
[[40  3]
 [ 1 70]]



Accuracy:

The test accuracy (96.5%) is higher than the training accuracy (93.6%).
This means the model is not overfitted and performs better on the new data.
_____________________________________
Recall:

Test Recall = 98.6% -> Only 1.4% of malignant patients were not detected. But overall it is excellent.
_____________________________________
Precition:

Test Precision = 95.9% -> This means that when the model says "you have cancer", it is correct 96% of the time! So only 4% False Positives (false alarms).
_____________________________________
F1_score:

F1 test = 97.2% -> Perfect balance between Precision and Recall.
_____________________________________
Cofusion matrix test:

[TN = 40  FP = 3]

[FN = 1   TP = 70]

40 people: were healthy and the model correctly said "healthy"

3 people: were healthy but the model incorrectly said "sick"

1 person: were sick but the model incorrectly said "healthy"

70 people: were sick and the model correctly said "sick"